### PR TITLE
#1802 left/right trim

### PR DESCRIPTION
**Pull request by antonydandrea**
It was said in this thread that there was a preference of adding parameters to the trim filter in order to implement this.

So this feature works like:

trim(character_mask, mode), where mode is either "both", "left" or "right". It is by default "both".

This is my first pull request of this project, so I apologise in advance if something is wrong.

### CODE REVIEW

File: doc/filters/trim.rst [hunk 1]
```diff
@@ -14,6 +14,10 @@ and end of a string:
 
     {# outputs '  I like Twig' #}
 
+    {{ '  I like Twig.  '|rtrim('', 'left') }}
```
Comment (SpacePossum) [inline doc/filters/trim.rst]: `rtrim` is not added as filter in this PR, so I guess maybe a typo?

Comment (antonydandrea) [inline doc/filters/trim.rst]: Yes that is a typo. I have updated with a fix.

File: doc/filters/trim.rst [hunk 2]
```diff
@@ -22,5 +26,6 @@ Arguments
 ---------
 
 * ``character_mask``: The characters to strip
-
+* ``mode``: The default is to strip from the start and the end ('both'), but 'left'
+and 'right' will strip from either the left side or right side only.
```
Comment (xabbuh) [inline doc/filters/trim.rst]: I think you should keep a blank line before the next one.

Comment (xabbuh) [inline doc/filters/trim.rst]: And I would enclose `both`, `left` and `right` in double backticks to get proper syntax highlighting in the rendered docs.

File: lib/Twig/Extension/Core.php [hunk 3]
```diff
@@ -852,6 +852,34 @@ function twig_in_filter($value, $compare)
...
+ * 
+ * @param string $string
+ * @param string $characterMask
+ * @param string $mode
+ * @return string
+ */
+function twig_trim_filter($string, $characterMask = null, $mode = 'both')
+{
+    if (empty($characterMask)) {
+        $characterMask = " \t\n\r\0\x0B";
```
Comment (xabbuh) [inline lib/Twig/Extension/Core.php]: Why not setting this as the parameter's default value instead?

File: lib/Twig/Extension/Core.php [hunk 4]
```diff
@@ -852,6 +852,34 @@ function twig_in_filter($value, $compare)
...
+    }
+    switch ($mode) {
+        case 'left':
+            $trimmed = ltrim($string, $characterMask);
```
Comment (xabbuh) [inline lib/Twig/Extension/Core.php]: You can just return here instead of introducing the `$trimmed` variable (same below).

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: Is it not better practice to have only one "Return" in a function rather than multiple?

File: lib/Twig/Extension/Core.php (hunk 3 above)
Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: If someone passes "empty string" instead of null as the first parameter, I believe it will override the default mask. So this catches all of those possibilities.

Comment (jvasseur) [inline lib/Twig/Extension/Core.php]: An empty string should not make the function function work like passing no parameter/passing null.

And this would be a BC break since currently `|trim('')` doesn't trim anything.

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: Understood. How should we handle this first parameter when people want to have the second parameter but use only the default mask?

Comment (jvasseur) [inline lib/Twig/Extension/Core.php]: Either `|trim(mode='left')` or `|trim(null, 'left')`.

File: lib/Twig/Extension/Core.php [hunk 5]
```diff
@@ -860,8 +860,11 @@ function twig_in_filter($value, $compare)
  * @param string $mode
  * @return string
  */
-function twig_trim_filter($string, $characterMask = " \t\n\r\0\x0B", $mode = 'both')
+function twig_trim_filter($string, $characterMask = null, $mode = 'both')
 {
+    if (empty($characterMask)) {
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: I think a check against `null` is better, as this way you cannot set `$characterMask` to a single space (' ' is empty).

File: lib/Twig/Extension/Core.php [hunk 6]
```diff
@@ -852,6 +852,34 @@ function twig_in_filter($value, $compare)
...
 
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $characterMask
+ * @param string $mode
+ * @return string
+ */
+function twig_trim_filter($string, $characterMask = null, $mode = 'both')
```
Comment (javiereguiluz) [inline lib/Twig/Extension/Core.php]: If I'm right, in English they say *"trim from [left/right/both sides]"* ... so maybe we could rename the `$mode` parameter to either `from` or `side`:

```twig
{{ '...'|trim(mode='both') }}
{{ '...'|trim(mode='left') }}
{{ '...'|trim(mode='right') }}

{{ '...'|trim(from='both') }}
{{ '...'|trim(from='left') }}
{{ '...'|trim(from='right') }}

{{ '...'|trim(side='both') }}
{{ '...'|trim(side='left') }}
{{ '...'|trim(side='right') }}
```

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: Also throwing out there, PHP's "str_pad" uses the word "_type_" for declaring left and right.

File: lib/Twig/Extension/Core.php [hunk 7]
```diff
@@ -852,6 +852,33 @@ function twig_in_filter($value, $compare)
...
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $character_mask
+ * @param string $mode
+ * @return string
+ */
+function twig_trim_filter($string, $character_mask = null, $type = 'both')
+{
+    if (is_null($character_mask)) {
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: please use `null ===`  (slightly faster and cannot be overridden)

File: lib/Twig/Extension/Core.php [hunk 8]
```diff
@@ -852,6 +852,33 @@ function twig_in_filter($value, $compare)
...
+        $character_mask = " \t\n\r\0\x0B";
+    }
+    switch ($type) {
+        case 'left':
+            return ltrim($string, $character_mask);
+            break;
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: these `breaks` now have no function, not sure if these should stay?

File: lib/Twig/Extension/Core.php [hunk 9]
```diff
@@ -852,6 +852,30 @@ function twig_in_filter($value, $compare)
...
+    if (null === $character_mask) {
+        $character_mask = " \t\n\r\0\x0B";
+    }
+    switch ($type) {
+        case 'left':
+            return ltrim($string, $character_mask);
+        case 'right':
+            return rtrim($string, $character_mask);
+        case 'both':
+        default:
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: Wouldn't it be better to throw an exception here instead of fallbacking to `both`, which is already the default value of `$type` parameter anyway?

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: @julienfalque I did think this, but I wasn't sure which Exception. Is there a specific Twig Exception you would throw?

File: lib/Twig/Extension/Core.php [hunk 10]
```diff
@@ -852,6 +852,30 @@ function twig_in_filter($value, $compare)
...
 
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $character_mask
+ * @param string $mode
+ * @return string
+ */
+function twig_trim_filter($string, $character_mask = null, $type = 'both')
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: Should be `$characterMask`.

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: @julienfalque would that work for named parameters? As the documentation has "character_mask".

File: lib/Twig/Extension/Core.php (hunk 9 above)
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: I guess `Twig_Error_Runtime` is fine.

File: lib/Twig/Extension/Core.php (hunk 10 above)
Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: @julienfalque answered my own question: it does work.

File: doc/filters/trim.rst [hunk 11]
```diff
@@ -14,6 +14,10 @@ and end of a string:
 
     {# outputs '  I like Twig' #}
 
+    {{ '  I like Twig.  '|trim(type='left') }}
+
+    {# outputs '  I like Twig. #}
```
Comment (lyrixx) [inline doc/filters/trim.rst]: It looks like a `'` is missing.

File: lib/Twig/Extension/Core.php (hunk 10 above)
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: Great, I forgot about named arguments so I learnt something :)

File: lib/Twig/Extension/Core.php [hunk 12]
```diff
@@ -852,6 +852,30 @@ function twig_in_filter($value, $compare)
     return false;
 }
 
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $character_mask
+ * @param string $mode
+ * @return string
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: Missing empty line before `@return` tag.

File: lib/Twig/Extension/Core.php [hunk 13]
```diff
@@ -852,6 +852,30 @@ function twig_in_filter($value, $compare)
     return false;
 }
 
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $character_mask
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: Should be `$characterMask`.

File: doc/filters/trim.rst [hunk 14]
```diff
@@ -23,4 +27,7 @@ Arguments
 
 * ``character_mask``: The characters to strip
 
+* ``type``: The default is to strip from the start and the end (`both`), but `left`
+and `right` will strip from either the left side or right side only.
```
Comment (julienfalque) [inline doc/filters/trim.rst]: This line should be indented with two spaces so it is rendered as part of the bullet point instead of a new line. See https://github.com/antonydandrea/Twig/blob/2.x_ltrim_rtrim/doc/filters/trim.rst.

File: lib/Twig/Extension/Core.php [hunk 15]
```diff
@@ -852,6 +852,34 @@ function twig_in_filter($value, $compare)
...
+    }
+    switch ($type) {
+        case 'left':
+            return ltrim($string, $characterMask);
+        case 'right':
+            return rtrim($string, $characterMask);
+        case 'both':
+            return trim($string, $characterMask);
+        default:
+            throw new Twig_Error_Runtime(sprintf('trim type must be "left", "right" or "both". Got %s.', $type));
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: "Trimming type"?

Btw, I think this will break if `$type` is passed something that cannot be converted to a string, e.g. an array.

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: Yes it would. Is there any value showing the given type?

Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: The function receives the raw value, so you can use `gettype()`, `is_array()` or whatever you want to use.

Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]:
```php
throw new Twig_Error_Runtime(sprintf('trim type must be "left", "right" or "both", got "%s".', is_object($type) ? get_class($type) : (gettype($type).'#'.$type)));
```
(this will fail `resources` though)

On a side note, as `both` is prop. used the most, please make it the first check of the `switch`

File: lib/Twig/Extension/Core.php [hunk 16]
```diff
@@ -852,6 +852,39 @@ function twig_in_filter($value, $compare)
...
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $characterMask
+ * @param string $mode
+ *
+ * @return string
+ *
+ * @throws Twig_Error_Runtime When an type is not a string.
+ * @throws Twig_Error_Runtime When an invalid type is used (not 'left', 'right' or 'both')
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: can be merged into one line

File: lib/Twig/Extension/Core.php [hunk 17]
```diff
@@ -852,6 +852,39 @@ function twig_in_filter($value, $compare)
...
+ */
+function twig_trim_filter($string, $characterMask = null, $type = 'both')
+{
+    if (!is_string($type)) {
+        throw new Twig_Error_Runtime(sprintf('Trimming type must be a string but got %s.', gettype($type)));
+    }
+    if (!in_array($type, ['both', 'left', 'right'])) {
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: I would throw the exception in the `default` of the switch, as most of the time this will be false it is better for speed to not test it every time on forhand

File: lib/Twig/Extension/Core.php [hunk 18]
```diff
@@ -852,6 +852,39 @@ function twig_in_filter($value, $compare)
...
+        throw new Twig_Error_Runtime(sprintf('Trimming type must be "left", "right" or "both" but got %s.', $type));
+    }
+    if (null === $characterMask) {
+        $characterMask = " \t\n\r\0\x0B";
+    }
```
Comment (SpacePossum) [inline lib/Twig/Extension/Core.php]: One last CS thingy, please add a blank line after a control structure.
Beside these few minors this looks good to me, thanks for adding this, much appreciated  :+1:  :)

File: lib/Twig/Extension/Core.php [hunk 19]
```diff
@@ -852,6 +852,41 @@ function twig_in_filter($value, $compare)
...
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $characterMask
+ * @param string $mode
+ *
+ * @return string
+ *
+ * @throws Twig_Error_Runtime When an type is not a string or when an invalid trimming type is used (not 'left', 'right' or 'both')
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]:
```diff
-When an type
+When trimming type
```

IMO a shorter sentence like _"When an invalid trimming type is used"_ would be enough, but I'm fine with this one.

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: You're right, there is a typo there also, and it is a bit wordy

**Comment by SpacePossum** [comment]
:+1: shippit :)

**Comment by antonydandrea** [comment]
Thanks everyone for being so welcoming to a newbie and for helping make the PR meet the coding standards of the project. I hope to contribute again in the future!

File: lib/Twig/Extension/Core.php [hunk 20]
```diff
@@ -852,6 +852,41 @@ function twig_in_filter($value, $compare)
...
+    switch ($type) {
+        case 'both':
+            return trim($string, $characterMask);
+        case 'left':
+            return ltrim($string, $characterMask);
+        case 'right':
+            return rtrim($string, $characterMask);
+        default:
+            if (!is_string($type)) {
+                throw new Twig_Error_Runtime(sprintf('Trimming type must be a string but got %s.', gettype($type)));
```
Comment (fabpot) [inline lib/Twig/Extension/Core.php]: I would remove this test a we are not doing that anywhere else and instead, change the condition that follows to: `if (!is_array($type) || !in_array($type, array('both', 'left', 'right'))) {`

File: doc/filters/trim.rst [hunk 21]
```diff
@@ -14,13 +14,26 @@ and end of a string:
...
+    {# outputs 'I like Twig.  ' #}
+
+    {{ '  I like Twig.  '|trim(' ', 'right') }}
+
+    {# outputs '  I like Twig.' #}
+
 .. note::
 
-    Internally, Twig uses the PHP `trim`_ function.
+    Internally, Twig uses the PHP `trim`_, `ltrim`_ and `rtrim`_ functions.
```
Comment (fabpot) [inline doc/filters/trim.rst]: missing comma before `and`.

File: lib/Twig/Extension/Core.php [hunk 22]
```diff
@@ -852,6 +852,41 @@ function twig_in_filter($value, $compare)
     return false;
 }
 
+/** 
+ * Returns a trimmed string.
+ * 
+ * @param string $string
+ * @param string $characterMask
+ * @param string $type
```
Comment (fabpot) [inline lib/Twig/Extension/Core.php]: The three `@param` can be removed

File: doc/filters/trim.rst [hunk 23]
```diff
@@ -14,13 +14,26 @@ and end of a string:
...
 
 Arguments
 ---------
 
 * ``character_mask``: The characters to strip
 
+* ``type``: The default is to strip from the start and the end (`both`), but `left`
+  and `right` will strip from either the left side or right side only.
```
Comment (fabpot) [inline doc/filters/trim.rst]: Can you remove the dot at the end?

File: doc/filters/trim.rst [hunk 24]
```diff
@@ -14,13 +14,26 @@ and end of a string:
 
     {# outputs '  I like Twig' #}
 
+    {{ '  I like Twig.  '|trim(type='left') }}
```
Comment (fabpot) [inline doc/filters/trim.rst]: You should add a `versionadded` tag at the beginning of the file (like done in `doc/filters/convert_encoding.rst` for instance):

```
.. versionadded:: 1.32
    The ``type`` argument was added in Twig 1.32.
```

**Comment by fabpot** [comment]
This feature should be added in the 1.x branch instead. You can change the target here and rebase your changes (if you're not sure about this, keep it as is and I will take care of that when merging).

Can you add a note in the CHANGELOG file as well about this new addition?

**Comment by antonydandrea** [comment]
@fabpot struggling with the rebase to be honest :-S

**Comment by fabpot** [comment]
@antonydandrea No worries, I will take care of that.

I'm not a big fan of `type` as the word does not say much about the feature. What about trying to be more precise like `position` or `side`, or ...? Any other ideas?

**Comment by robfrawley** [comment]
I rather like `side`, as it is short (if using named arguments, this can be important) and fairly precise. Good job on your first PR for `twig/twig` @antonydandrea!

**Comment by antonydandrea** [comment]
I think "side" is best. 

When I originally picked "mode" and then "type", I did so because I thought it kept consistent with the terminology in PHP. But I realise now, that Twig is suppose to abstract away from "developer speak".

I think "side" is the most natural way of describing this: for example, 
"I want to trim the spaces." 
"Which side? The left side? The right side? Or both sides?"

Other words I considered I think have other connotations: "position" is more to point to a specific point; "end" (which end do you want to trim), is too ambiguous because it can be used in the context of "start" and "end".

File: lib/Twig/Extension/Core.php [hunk 25]
```diff
@@ -852,6 +852,33 @@ function twig_in_filter($value, $compare)
...
+
+    switch ($side) {
+        case 'both':
+            return trim($string, $characterMask);
+        case 'left':
+            return ltrim($string, $characterMask);
+        case 'right':
+            return rtrim($string, $characterMask);
+        default:
+            if (!is_string($side) || !in_array($side, ['both', 'left', 'right'])) {
```
Comment (julienfalque) [inline lib/Twig/Extension/Core.php]: I think this `if` is unnecessary: if we are in the `default` case, then we already know that `$side` is not a string or has an invalid value.

Comment (antonydandrea) [inline lib/Twig/Extension/Core.php]: Ah true. This evolved from when I was capturing both conditions separately to give two different errors - which is no longer the case.

**Comment by fabpot** [comment]
Thank you @antonydandrea.
